File: EldersFirebaseRemoteConfig/RemoteConfig+Defaults.m
//
//  RemoteConfig+Defaults.m
//  EldersFirebaseRemoteConfig
//
//  Created by Milen Halachev on 18.03.21.
//

#import "RemoteConfig+Defaults.h"
#import <objc/runtime.h>
#import <EldersFirebaseRemoteConfig/EldersFirebaseRemoteConfig-Swift.h>

@implementation FIRRemoteConfig (Defaults)

+ (void)load {
    
    Method old = class_getInstanceMethod([self class], @selector(setDefaults:));
    Method new = class_getInstanceMethod([self class], @selector(swizzled_setDefaults:));
    method_exchangeImplementations(old, new);
}

- (void)swizzled_setDefaults:(nullable NSDictionary<NSString*, NSObject *>*)defaultConfig {
    
    //inject custom librabry defaults here
    NSMutableDictionary<NSString*, NSObject *> *defaults = [[[self class] eldersDefaults] mutableCopy];
    [defaults addEntriesFromDictionary:defaultConfig];

    [self swizzled_setDefaults:defaults];
}

@end

File: EldersFirebaseRemoteConfig/RemoteConfig+Defaults.swift
//
//  RemoteConfig+Defaults.swift
//  EldersFirebaseRemoteConfig
//
//  Created by Milen Halachev on 18.03.21.
//

import Foundation
import FirebaseRemoteConfig

extension RemoteConfig {
    
    class func eldersDefaults() -> [Key: NSObject] {

        var eldersDefaults: [Key: NSObject] = [:]

        if ProcessInfo.processInfo.arguments.contains("UNIT_TESTS") {

            eldersDefaults["eldersDefaultsTestValue1"] = "da ve" as NSString
            eldersDefaults["eldersDefaultsTestValue2"] = NSNumber(7)
        }

        return eldersDefaults
    }
    
    @objc public class func eldersDefaults() -> [String: NSObject]  {
        
        let eldersDefaults: [Key: NSObject] = self.eldersDefaults()
        return eldersDefaults.map()
    }
}
